### PR TITLE
Hw4 task01

### PR DESCRIPTION
**Pull request by mirmozavr**
Open text, read the first line

### CODE REVIEW

File: hw4/hw4_t01.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+def read_magic_number(path: str) -> bool:
+    try:
+        with open(path, "r", encoding="utf-8") as text:
+            first_line = text.readline().strip()
+            if first_line.isdigit() and 1 <= int(first_line) < 3:
+                return True
+            return False
+    except Exception:
+        raise ValueError("Error text")
```
Comment (Cjkjvfnby) [inline hw4/hw4_t01.py]: Question text

File: hw4/hw4_t01.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    finally:
+        rmtree(os.path.dirname(__file__) + "/__pycache__", ignore_errors=True)
```
Comment (Cjkjvfnby) [inline hw4/hw4_t01.py]: > all temporary files are removed after a test run

It is not about these files. Only about files you have created. 

After test run means only after the test run, not during the function call.

File: hw4/tests/test_hw4_t01.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+def test_valid_1():
+    assert read_magic_number(os.path.dirname(__file__) + "/data1.txt") is True
+
+
+def test_valid_2():
+    assert read_magic_number(os.path.dirname(__file__) + "/data2.txt") is True
+
+
+def test_not_valid_3():
+    assert read_magic_number(os.path.dirname(__file__) + "/data3.txt") is False
```
Comment (Cjkjvfnby) [inline hw4/tests/test_hw4_t01.py]: All `data?.txt` looks the same for me. But some of them are True and others False.  Please make names more descriptive.

File: hw4/hw4_t01.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+Definition of done:
+ - function is created
+ - function is properly formatted
+ - function has positive and negative tests
+ - all temporary files are removed after test run
+
+You will learn:
+ - how to test Exceptional cases
+ - how to clean up after tests
```
Comment (Cjkjvfnby) [inline hw4/hw4_t01.py:19]: Have you learned it?

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]
I have updated the task description to clarify what I expect a generation of the test files inside the tests.

File: hw4/hw4_t01.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+file_path = os.path.dirname(__file__) + "/tests/data221.txt"
+
+
+def read_magic_number(path: str) -> bool:
+    try:
+        with open(path, "r", encoding="utf-8") as text:
+            first_line = text.readline().strip()
+            if first_line.isdigit() and 1 <= int(first_line) < 3:
+                return True
+            return False
```
Comment (Cjkjvfnby) [inline hw4/hw4_t01.py]: Can you simplify it?

File: hw4/tests/test_hw4_t01.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+"""
+    )
+    assert read_magic_number(directory / "test_data.txt") is False
+
+
+def test_file_doesnt_exist_value_error(tmp_path):
+    directory = tmp_path / "sub_directory"
+    directory.mkdir()
+    with pytest.raises(ValueError):  # noqa: PT011
+        read_magic_number(directory / "missing_file.txt")
```
Comment (Cjkjvfnby) [inline hw4/tests/test_hw4_t01.py]: Why do you need a directory here?

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]
Some cosmetic comments.  

Please describe how cleanup works.

**Comment by mirmozavr** [comment]
Quote from docs.pytest.org
Temporary directories are by default created as sub-directories of the system temporary directory. The base name will be pytest-NUM where NUM will be incremented with each test run. Moreover, entries older than 3 temporary directories will be removed.
They are removed indeed, I checked just to be sure

**Review by Cjkjvfnby — APPROVED** [review]

**Pull request merged**